kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopCredentialsManager.scala

```diff
@@ -0,0 +1,242 @@
+ * </li>
+ * <li>
+ *   If sending succeeds, set session credentials epoch to [[CredentialsRef]] epoch. Else,
+ *   record the exception and return.
+ * </li>
+ * </ol>
+ *
+ * @note Session credentials epochs are created in session scope and should be removed using
+ *       [[removeSessionCredentialsEpoch]] when session closes.
+ */
+class HadoopCredentialsManager private (name: String) extends AbstractService(name)
+    with Logging {
+
+  def this() = this(classOf[HadoopCredentialsManager].getSimpleName)
+
+  private val userCredentialsRefMap = new ConcurrentHashMap[String, CredentialsRef]()
+  private val sessionCredentialsEpochMap = new ConcurrentHashMap[String, Long]()
+
+  private var providers: Map[String, HadoopDelegationTokenProvider] = _
+  private var renewalInterval: Long = _
+  private var renewalRetryWait: Long = _
+  private var renewalExecutor: ScheduledExecutorService = _
+  private var hadoopConf: Configuration = _
+
+  override def initialize(conf: KyuubiConf): Unit = {
+    hadoopConf = KyuubiHadoopUtils.newHadoopConf(conf)
+    providers = HadoopCredentialsManager.loadProviders(conf)
+      .filter { case (_, provider) =>
+        val required = provider.delegationTokensRequired(hadoopConf, conf)
+        if (!required) {
+          info(s"Service ${provider.serviceName} does not require a token." +
+            s" Check your configuration to see if security is disabled or not.")
+        }
+        required
+      }
+    info("Using the following builtin delegation token providers: " +
+      s"${providers.keys.mkString(", ")}.")
+
+    renewalInterval = conf.get(CREDENTIALS_RENEWAL_INTERVAL)
+    renewalRetryWait = conf.get(CREDENTIALS_RENEWAL_RETRY_WAIT)
+    super.initialize(conf)
+  }
+
+  override def start(): Unit = {
+    renewalExecutor =
+      ThreadUtils.newDaemonSingleThreadScheduledExecutor("Delegation Token Renewal Thread")
+    super.start()
+  }
+
+  override def stop(): Unit = {
+    if (renewalExecutor != null) {
+      renewalExecutor.shutdownNow()
+      try {
+        renewalExecutor.awaitTermination(10, TimeUnit.SECONDS)
+      } catch {
+        case _: InterruptedException =>
+      }
+    }
+    super.stop()
+  }
+
+  /**
+   * Send credentials to SQL engine which the specified session is talking to if
+   * [[HadoopCredentialsManager]] has a newer credentials.
+   *
+   * @param sessionId Specify the session which is talking with SQL engine
+   * @param appUser  User identity that the SQL engine uses.
+   * @param send     Function to send encoded credentials to SQL engine
+   */
+  def sendCredentialsIfNeeded(
+      sessionId: String,
+      appUser: String,
+      send: String => Unit): Unit = {
+    require(renewalExecutor != null, "renewalExecutor should be initialized")
+
+    val userRef = getOrCreateUserCredentialsRef(appUser)
+    val sessionEpoch = getSessionCredentialsEpoch(sessionId)
+
+    if (userRef.getEpoch != sessionEpoch) {
```

Comment:
so there are two cases the user epoch can be lager than session epoch
* the new created user cred
* we failed to send created before

How about use `if (userRef.getEpoch > sessionEpoch)` ?

Reply:
done